plugins/bnr-dp/fu-bnr-dp-device.c

```diff
@@ -0,0 +1,748 @@
+
+	fu_struct_bnr_dp_aux_command_set_module_number(command, module_number);
+	fu_struct_bnr_dp_aux_command_set_opcode(command, opcode);
+
+	if (!fu_struct_bnr_dp_aux_request_set_command(request, command, error))
+		return NULL;
+	fu_struct_bnr_dp_aux_request_set_data_len(request, data_len);
+	fu_struct_bnr_dp_aux_request_set_offset(request, offset);
+
+	return g_steal_pointer(&request);
+}
+
+/* evaluate the status from a response from the controller into an appropriate bool/GError */
+static gboolean
+fu_bnr_dp_device_eval_result(const FuStructBnrDpAuxStatus *status, GError **error)
+{
+	guint8 error_byte = fu_struct_bnr_dp_aux_status_get_error(status);
+	guint8 error_code = error_byte & 0x0F;
+
+	if (error_byte & FU_BNR_DP_AUX_STATUS_FLAGS_ERROR) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_READ,
+			    "device command failed with error '%s'",
+			    fu_bnr_dp_aux_error_to_string(error_code) ?: "(invalid error code)");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_bnr_dp_device_is_busy(const FuStructBnrDpAuxStatus *status)
+{
+	guint8 error_byte = fu_struct_bnr_dp_aux_status_get_error(status);
+
+	return error_byte & FU_BNR_DP_AUX_STATUS_FLAGS_BUSY;
+}
+
+/* Write a single `request` and some optional data to the device. */
+static gboolean
+fu_bnr_dp_device_write_request(FuBnrDpDevice *self,
+			       const FuStructBnrDpAuxRequest *request,
+			       const guint8 *buf,
+			       gsize bufsz,
+			       GError **error)
+{
+	guint8 checksum =
+	    fu_bnr_dp_device_xor_checksum(FU_BNR_DP_CHECKSUM_INIT_TX, request->data, request->len);
+	g_autoptr(FuStructBnrDpAuxTxHeader) header = fu_struct_bnr_dp_aux_tx_header_new();
+
+	if (!fu_struct_bnr_dp_aux_tx_header_set_request(header, request, error))
+		return FALSE;
+
+	/* write optional data */
+	if (buf != NULL && bufsz > 0) {
+		if (!fu_dpaux_device_write(FU_DPAUX_DEVICE(self),
+					   FU_BNR_DP_DEVICE_DATA_OFFSET,
+					   buf,
+					   bufsz,
+					   FU_BNR_DP_DEVICE_DPAUX_TIMEOUT_MSEC,
+					   error))
+			return FALSE;
+
+		checksum = fu_bnr_dp_device_xor_checksum(checksum, buf, bufsz);
+	}
+
+	fu_struct_bnr_dp_aux_tx_header_set_checksum(header, checksum);
+
+	/* write header to kick off processing by the device */
+	return fu_dpaux_device_write(FU_DPAUX_DEVICE(self),
+				     FU_BNR_DP_DEVICE_HEADER_OFFSET,
+				     header->data,
+				     header->len,
+				     FU_BNR_DP_DEVICE_DPAUX_TIMEOUT_MSEC,
+				     error);
+}
+
+/* read a single `response` and some optional data from the device after a finished command. reading
+ * the full 7 byte header from the header offset returns a different structure than when reading
+ * only 2 bytes */
+static gboolean
+fu_bnr_dp_device_read_response(FuBnrDpDevice *self, GByteArray *data, GError **error)
+{
+	guint8 actual_checksum;
+	guint8 tmp[FU_STRUCT_BNR_DP_AUX_RX_HEADER_SIZE] = {0};
+	g_autoptr(FuStructBnrDpAuxRxHeader) header = NULL;
+	g_autoptr(FuStructBnrDpAuxResponse) response = NULL;
```

Comment:
it's a trivial style thing, but rustgen-generated structs usually have a `st_` prefix -- e.g. `st_header`